src/protagonist/Portal/Features/Batches/Requests/ParseCsv.cs

```diff
@@ -0,0 +1,198 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Security.Claims;
+using System.Text.Json.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using API.Client;
+using DLCS.HydraModel;
+using Hydra.Collections;
+using MediatR;
+using Microsoft.AspNetCore.Http;
+using CsvHelper;
+using CsvHelper.Configuration.Attributes;
+using CsvHelper.TypeConversion;
+using DLCS.Web.Auth;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Portal.Settings;
+using MissingFieldException = CsvHelper.MissingFieldException;
+
+namespace Portal.Features.Batches.Requests;
+
+public class ParseCsv : IRequest<ParseCsvResult>
+{
+    public IFormFile File { get; set; }
+}
+
+public class ParseCsvResult
+{
+    [JsonIgnore] public bool IsSuccess { get; set; }
+
+    public string[]? Errors { get; set; }
+
+    public static ParseCsvResult Failure(string[] errors)
+    {
+        return new ParseCsvResult { IsSuccess = false, Errors = errors };
+    }
+    
+    public static ParseCsvResult Failure(string error)
+    {
+        return new ParseCsvResult { IsSuccess = false, Errors = new []{ error } };
+    }
+    
+    public static ParseCsvResult Success()
+    {
+        return new ParseCsvResult { IsSuccess = true };
+    }
+}
+
+public class ParseCsvHandler : IRequestHandler<ParseCsv, ParseCsvResult>
+{
+    private readonly IDlcsClient dlcsClient;
+    private readonly ILogger<ParseCsvHandler> logger;
+    private readonly int customerId;
+    private readonly int maxBatchSize;
+    
+    public ParseCsvHandler(IDlcsClient dlcsClient, IOptions<PortalSettings> portalSettings, ILogger<ParseCsvHandler> logger, ClaimsPrincipal currentUser)
+    {
+        this.dlcsClient = dlcsClient;
+        this.logger = logger;
+        customerId = (currentUser.GetCustomerId() ?? -1);
+        maxBatchSize = portalSettings.Value.MaxBatchSize;
+    }
+
+    public async Task<ParseCsvResult> Handle(ParseCsv request, CancellationToken cancellationToken)
+    {
+        var distinctRows = new Dictionary<int, Image>();
+        var readErrors = new List<string>();
+        
+        using (var stream = new StreamReader(request.File.OpenReadStream()))
+        using (var csv = new CsvReader(stream, CultureInfo.InvariantCulture))
+        {
+            csv.Read();
+            csv.ReadHeader();
+            while (csv.Read())
```

Comment:
Rider is showing that there's an async overload of `csv.Read()` that is probably worth using, particularly for larger files